Implement field type `media`

oandregal opened the pull request:
Follow-up to https://github.com/wordpress/gutenberg/pull/67278#discussion_r1905408381 and https://github.com/Automattic/wp-calypso/pull/103149#issuecomment-2853850451

## Proposed Changes

Implements the `media` field type.

## Why are these changes being made?

DataViews declares a field type `media`, but there's no implementation for that field type.

So far, the type was used to provide an affordance to switch the media field ([see](https://github.com/wordpress/gutenberg/pull/67278#discussion_r1905408381)). With this PR, fields that set a `media` type would also have sorting disabled by default — though they can override it.

## Testing Instructions

- In the storybook (`yarn workspace @automattic/dataviews storybook:start`):
  - go to the DataViews default story and verify that the "Image" field is not sortable either.
  - in this case, there's only one media type field, so the "media switcher" is not available in the view config.

oandregal committed:
Make preview and icon.ico of type media

oandregal committed:
Add field type media

oandregal committed:
Field types can now provide enableSorting

oandregal commented:
Before:

https://github.com/user-attachments/assets/d0fe9415-3cc3-4528-9f75-3342e7f846bd

After:

https://github.com/user-attachments/assets/f65ce6fc-07ef-4b09-b46d-39a349a60326

oandregal committed:
Update changelog

youknowriad commented:
What would be cool is a way to "override" bundled types. For instance, it would be cool for WP (editor or edit-site or else) to provide a definition for the "edit" part (with uploads, and dependency to core-data). (Just a random thought for now, not necessarily something to act on)

oandregal commented:
Yeah, `Edit` for media is bound to the environment, so it's difficult to provide a good default for it. I also think `media` is too broad that we cannot even provide a good default for `render`: it feels like providing a `alphanumeric` type instead of `integer`, `text`, etc.

youknowriad commented:
Should we have a bundled "media" type? What's the reason for its existence?

oandregal commented:
> Should we have a bundled "media" type? What's the reason for its existence?

It was added at https://github.com/WordPress/gutenberg/pull/67278 to support this: let users switch the "media field" in the view config.

My view is that if a fieldtype exists, it should be implemented, otherwise we'll run into bugs — that's what this PR does. I also think it's incomplete, and we need to iterate.

youknowriad commented:
> It was added at https://github.com/WordPress/gutenberg/pull/67278 to support this: let users switch the "media field" in the view config.

I wonder if this should be a "type" or some kind of "support" flag.

oandregal commented:
> I wonder if this should be a "type" or some kind of "support" flag.

Yeah. We had the same conversation a few months ago 😅 https://github.com/WordPress/gutenberg/pull/67278#discussion_r1871652925

ntsekouras commented:
I'm also not sure if we should add the type If we can't have a proper `default` render. 

Maybe it's better to revisit the preview functionality without the `type:media` like Riad also suggests? A flag might be a better fit after all, since even now it **shouldn't** mean that every `image/media` in a list should be a candidate to update the preview(mediaField).

oandregal committed:
Revert "Make preview and icon.ico of type media"

This reverts commit 63c4416ad8263721b1c538a20118400670a1f2b5.

oandregal commented:
@youknowriad @ntsekouras Let me rephrase, I haven't done a great job at explaining.

We already have the type `media`, it was introduced [in Gutenberg](https://github.com/WordPress/gutenberg/pull/67278#discussion_r1905408381). However, we didn't implement it and this is an issue because we never followed up to do it. What happens in this situation?

```js
{
  id: 'image',
  label: 'Image',
  type: 'media'
}
```

Take sorting, for example. How sorting would work for a field type that hasn't defined sorting? Because type is not mandatory, it fallbacks to the untyped behavior, and _it sorts field of type media as if they were text_. This is a bug we should fix. I strongly believe any field type should provide an implementation, otherwise, we'll run into subtle bugs like the one I described. Alternatively, we could add fallbacks for types that don't provide an implementation in various places, but I think this adds unnecessary code complexity — it's simpler if we reason in terms of field types.

Additionally, this PR adds the `enableSorting` default, so field types can declare it, making the `media` type not sortable by default. This was how it was intended to work, and I think it's a nice thing to have. I can remove it if it's a blocker.

I've removed the changes to make sites fields, so that behavior remains untouched. **Would you agree to merge this PR as a bug fix to what I've described?** That's what I'm invested in to deploy with this PR. Happy to hear alternatives.

> Maybe it's better to revisit the preview functionality without the type:media like Riad also suggests? A flag might be a better fit after all, since even now it shouldn't mean that every image/media in a list should be a candidate to update the preview(mediaField).

Based on this comment, it looks like the issue is not with this PR but how the `media` field type behaves (that it adds all fields to the "preview" feature). I don't disagree with what you said, but it isn't solved by switching to a flag.

If we used the flag in the two existing sites' fields (media and icon), the two will still be available in the "preview" feature. The issue is that we're providing all fields for all views, but you only want the `icon.ico` to appear as the preview of the table, and `preview` as the preview of the grid. If we wanted to have a single media field per view type, we need to filter the fields depending on the view type — no matter we are using field type or a flag to detect them.

Anyway, as I mentioned, I removed the changes to sites, so this PR doesn't modify the dashboard behavior. This should be a separate follow-up/conversation.

youknowriad commented:
> If we used the flag in the two existing sites' fields (media and icon), the two will still be available in the "preview" feature. The issue is that we're providing all fields for all views, but you only want the icon.ico to appear as the preview of the table, and preview as the preview of the grid. If we wanted to have a single media field per view type, we need to filter the fields depending on the view type — no matter we are using field type or a flag to detect them.

Why don't we want the "icon" to be used as a "preview" for grid? (as an option I mean)

youknowriad commented:
It's also not clear to me how this related to the issue of using the "type" or "flag". It's completely independent. What I was saying is that there's no clear way to know what a "media" is, we can't define an order, we can't define a "render", we can't define an "edit", so it might not really make sense as a field type at all (making this PR useless)

oandregal commented:
> It's also not clear to me how this related to the issue of using the "type" or "flag". It's completely independent. What I was saying is that there's no clear way to know what a "media" is, we can't define an order, we can't define a "render", we can't define an "edit", so it might not really make sense as a field type at all (making this PR useless)

What I said is that I don't disagree with that, but we can't remove the field type if we want to stay back compatible with Gutenberg — hence the bug I mentioned. This is a PR to fix it.

I agree we need more nuance for media types, and we can explore that when/if we need.

youknowriad commented:
> What I said is that I don't disagree with that, but we can't remove the field type if we want to stay back compatible with Gutenberg — hence the bug I mentioned. This is a PR to fix it.

Why do we want to be back compatible with Gutenberg. DataViews is a private package even in Gutenberg, so breaking changes are allowed.

oandregal commented:
> Why do we want to be back compatible with Gutenberg. DataViews is a private package even in Gutenberg, so breaking changes are allowed.

I thought about this and let's say that we want to create more specific types (image, video, etc.). Until we have them, what do we do with the existing media type?

Even in that case, I still think we'd benefit from landing this PR (prevents bugs with media until better types are developed and it's eventually removed).

ntsekouras commented:
>Even in that case, I still think we'd benefit from landing this PR (prevents bugs with media until better types are developed and it's eventually removed).

If we remove the `media` type, we could also default to no sorting, searching etc.. for any field without a `type`, couldn't we?

youknowriad reviewed (approved):
My opinion is still to remove the field type and replace it with a flag in the field definition. But I don't mind if this is seen as a temporary measure.
